homeassistant/auth/__init__.py

```diff
@@ -479,6 +492,35 @@ async def async_remove_refresh_token(
         for revoke_callback in callbacks:
             revoke_callback()
 
+    async def _async_remove_expired_refresh_tokens(
+        self, _: datetime | None = None
+    ) -> None:
+        """Remove expired refresh tokens."""
+        now = dt_util.utcnow()
+        for token in self._store.async_get_refresh_tokens()[:]:
+            if (expire_at := token.expire_at) is not None and expire_at <= now:
+                await self.async_remove_refresh_token(token)
+        self.async_track_next_refresh_token_expiration()
+
+    @callback
+    def async_track_next_refresh_token_expiration(self) -> None:
+        """Initialise all token expiration scheduled tasks."""
+        next_expiration = dt_util.utcnow() + REFRESH_TOKEN_EXPIRATION
+        for token in self._store.async_get_refresh_tokens():
+            if (
+                expire_at := token.expire_at
+            ) is not None and expire_at < next_expiration:
+                next_expiration = expire_at
```

Comment:
Since we are going to iterate these every time it would be much faster to store time as `float` from `time.time()` ...

Reply:
That would also avoid all the conversion in `async_load`

Reply:
```diff
diff --git a/homeassistant/auth/__init__.py b/homeassistant/auth/__init__.py
index e160a45db25..c8bea02af95 100644
--- a/homeassistant/auth/__init__.py
+++ b/homeassistant/auth/__init__.py
@@ -76,7 +76,9 @@ async def auth_manager_from_config(
     for module in modules:
         module_hash[module.id] = module
 
-    return AuthManager(hass, store, provider_hash, module_hash)
+    manager = AuthManager(hass, store, provider_hash, module_hash)
+    manager.async_setup()
+    return manager
 
 
 class AuthManagerFlowManager(data_entry_flow.FlowManager):
@@ -162,9 +164,14 @@ class AuthManager:
         self._revoke_callbacks: dict[str, list[CALLBACK_TYPE]] = {}
         self._expire_callbacks: dict[str, CALLBACK_TYPE] = {}
         self._expire_callback: CALLBACK_TYPE | None = None
+        self._remove_expired_job = HassJob(self._async_remove_expired_refresh_tokens)
 
-        hass.async_add_job(self.async_track_next_refresh_token_expiration)
+    @callback
+    def async_setup(self) -> None:
+        """Set up the auth manager."""
+        hass = self.hass
         hass.async_add_shutdown_job(HassJob(self.async_cancel_expiration_schedule))
+        self.async_track_next_refresh_token_expiration()
 
     @property
     def auth_providers(self) -> list[AuthProvider]:
@@ -431,7 +438,7 @@ class AuthManager:
                 token_type = models.TOKEN_TYPE_NORMAL
 
         if token_type == models.TOKEN_TYPE_NORMAL:
-            expire_at = dt_util.utcnow() + REFRESH_TOKEN_EXPIRATION
+            expire_at = time.time() + REFRESH_TOKEN_EXPIRATION
         else:
             expire_at = None
 
@@ -496,7 +503,7 @@ class AuthManager:
         self, _: datetime | None = None
     ) -> None:
         """Remove expired refresh tokens."""
-        now = dt_util.utcnow()
+        now = time.time()
         for token in self._store.async_get_refresh_tokens()[:]:
             if (expire_at := token.expire_at) is not None and expire_at <= now:
                 await self.async_remove_refresh_token(token)
@@ -505,7 +512,7 @@ class AuthManager:
     @callback
     def async_track_next_refresh_token_expiration(self) -> None:
         """Initialise all token expiration scheduled tasks."""
-        next_expiration = dt_util.utcnow() + REFRESH_TOKEN_EXPIRATION
+        next_expiration = time.time() + REFRESH_TOKEN_EXPIRATION
         for token in self._store.async_get_refresh_tokens():
             if (
                 expire_at := token.expire_at
@@ -513,7 +520,7 @@ class AuthManager:
                 next_expiration = expire_at
 
         self._expire_callback = async_track_point_in_utc_time(
-            self.hass, self._async_remove_expired_refresh_tokens, next_expiration
+            self.hass, self._remove_expired_job, dt_util.utc_from_timestamp(next_expiration)
         )
 
     async def async_cancel_expiration_schedule(self) -> None:
diff --git a/homeassistant/auth/auth_store.py b/homeassistant/auth/auth_store.py
index 7ec9b30488b..ff3704cd94b 100644
--- a/homeassistant/auth/auth_store.py
+++ b/homeassistant/auth/auth_store.py
@@ -262,7 +262,7 @@ class AuthStore:
         refresh_token.last_used_ip = remote_ip
         if refresh_token.expire_at:
             refresh_token.expire_at = (
-                refresh_token.last_used_at + REFRESH_TOKEN_EXPIRATION
+                refresh_token.last_used_at.timestamp() + REFRESH_TOKEN_EXPIRATION
             )
         self._async_schedule_save()
 
@@ -432,12 +432,6 @@ class AuthStore:
             else:
                 last_used_at = None
 
-            # old refresh_token don't have expire_at (pre-2024.2)
-            if expire_at_str := rt_dict.get("expire_at"):
-                expire_at = dt_util.parse_datetime(expire_at_str)
-            else:
-                expire_at = None
-
             token = models.RefreshToken(
                 id=rt_dict["id"],
                 user=users[rt_dict["user_id"]],
@@ -454,7 +448,7 @@ class AuthStore:
                 jwt_key=rt_dict["jwt_key"],
                 last_used_at=last_used_at,
                 last_used_ip=rt_dict.get("last_used_ip"),
-                expire_at=expire_at,
+                expire_at=rt_dict.get("expire_at"),
                 version=rt_dict.get("version"),
             )
             if "credential_id" in rt_dict:
@@ -528,9 +522,7 @@ class AuthStore:
                 if refresh_token.last_used_at
                 else None,
                 "last_used_ip": refresh_token.last_used_ip,
-                "expire_at": refresh_token.expire_at.isoformat()
-                if refresh_token.expire_at
-                else None,
+                "expire_at": refresh_token.expire_at,
                 "credential_id": refresh_token.credential.id
                 if refresh_token.credential
                 else None,
diff --git a/homeassistant/auth/const.py b/homeassistant/auth/const.py
index 11d1297fd35..704f5d1d57c 100644
--- a/homeassistant/auth/const.py
+++ b/homeassistant/auth/const.py
@@ -3,7 +3,7 @@ from datetime import timedelta
 
 ACCESS_TOKEN_EXPIRATION = timedelta(minutes=30)
 MFA_SESSION_EXPIRATION = timedelta(minutes=5)
-REFRESH_TOKEN_EXPIRATION = timedelta(days=90)
+REFRESH_TOKEN_EXPIRATION = timedelta(days=90).total_seconds()
 
 GROUP_ID_ADMIN = "system-admin"
 GROUP_ID_USER = "system-users"
diff --git a/homeassistant/auth/models.py b/homeassistant/auth/models.py
index b56c9a911da..4cf94401478 100644
--- a/homeassistant/auth/models.py
+++ b/homeassistant/auth/models.py
@@ -117,7 +117,7 @@ class RefreshToken:
     last_used_at: datetime | None = attr.ib(default=None)
     last_used_ip: str | None = attr.ib(default=None)
 
-    expire_at: datetime | None = attr.ib(default=None)
+    expire_at: float | None = attr.ib(default=None)
 
     credential: Credentials | None = attr.ib(default=None)
 
diff --git a/tests/auth/test_init.py b/tests/auth/test_init.py
index 970ecef2bf0..494bfc440b9 100644
--- a/tests/auth/test_init.py
+++ b/tests/auth/test_init.py
@@ -590,7 +590,7 @@ async def test_remove_expired_refresh_token(hass: HomeAssistant) -> None:
     now = dt_util.utcnow()
     with freeze_time(now):
         refresh_token = await manager.async_create_refresh_token(user, CLIENT_ID)
-        assert refresh_token.expire_at == now + timedelta(days=90)
+        assert refresh_token.expire_at == now.timestamp() + timedelta(days=90).total_seconds()
 
     with freeze_time(now + timedelta(days=89, hours=23)):
         async_fire_time_changed(hass, now + timedelta(days=89, hours=23))
@@ -610,14 +610,14 @@ async def test_update_expire_at_refresh_token(hass: HomeAssistant) -> None:
     now = dt_util.utcnow()
     with freeze_time(now):
         refresh_token = await manager.async_create_refresh_token(user, CLIENT_ID)
-        assert refresh_token.expire_at == now + timedelta(days=90)
+        assert refresh_token.expire_at == now.timestamp() + timedelta(days=90).total_seconds()
 
     with freeze_time(now + timedelta(days=30)):
         async_fire_time_changed(hass, now + timedelta(days=30))
         await hass.async_block_till_done()
         assert manager.async_create_access_token(refresh_token)
         await hass.async_block_till_done()
-        assert refresh_token.expire_at == now + timedelta(days=30) + timedelta(days=90)
+        assert refresh_token.expire_at == now.timestamp() + timedelta(days=30).total_seconds() + timedelta(days=90).total_seconds()
 
 
 async def test_register_revoke_token_callback(mock_hass) -> None:
```

Reply:
The downside is that `last_used_at` and `created_at` are doing datetime objects

Reply:
For this case I think timestamps are a better fit

I try to use datetime for these cases:
- displaying to humans
- when I care about doing something on Wednesdays
- a specific, hour/minute

I try to use timestamps for these case:
- durations
- comparing/firing at a specific points in time

Reply:
i've applied your suggestion and reworked it to use timestamps instead of datetimes ->
 [use timestamp for expire_at bf04edf](https://github.com/home-assistant/core/pull/108428/commits/bf04edf9b54df0475c0b7bd9052d5b321954f5d7)

Reply:
i think reworking `last_used_at` and `created_at` (_maybe not `created_at` because it will never change_) to also be timestamps should go into another PR

Reply:
> i think reworking `last_used_at` and `created_at` (_maybe not `created_at` because it will never change_) to also be timestamps should go into another PR

100% agree. Definitely didn't intend to imply it should be done here